File: frontend/lib/models/attendance.dart
// lib/models/attendance.dart
class Attendance {
  final String id;
  final String employeeId;
  final String date;
  final DateTime checkInTime;
  final DateTime? checkOutTime;
  final String? checkInAddress;
  final String? checkOutAddress;
  final int? estimatedTimeToOffice;
  final int? distanceFromOffice;
  final double totalHours;
  final String status;

  Attendance({
    required this.id,
    required this.employeeId,
    required this.date,
    required this.checkInTime,
    this.checkOutTime,
    this.checkInAddress,
    this.checkOutAddress,
    this.estimatedTimeToOffice,
    this.distanceFromOffice,
    this.totalHours = 0,
    required this.status,
  });

  factory Attendance.fromJson(Map<String, dynamic> json) {
    return Attendance(
      id: json['_id'] ?? '',
      employeeId: json['employeeId'] ?? '',
      date: json['date'] ?? '',
      checkInTime: DateTime.parse(json['checkInTime']),
      checkOutTime: json['checkOutTime'] != null
          ? DateTime.parse(json['checkOutTime'])
          : null,
      checkInAddress: json['checkInAddress'],
      checkOutAddress: json['checkOutAddress'],
      estimatedTimeToOffice: json['estimatedTimeToOffice'],
      distanceFromOffice: json['distanceFromOffice'],
      totalHours: (json['totalHours'] ?? 0).toDouble(),
      status: json['status'] ?? 'CHECKED_IN',
    );
  }

  bool get isCheckedIn => status == 'CHECKED_IN';
  bool get isCheckedOut => status == 'CHECKED_OUT';
}

File: backend/models/Attendance.js
// models/Attendance.js
const mongoose = require("mongoose");

const AttendanceSchema = new mongoose.Schema(
  {
    employeeId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },
    date: {
      type: String,
      required: true,
    },
    checkInTime: {
      type: Date,
      required: true,
    },
    checkInLocation: {
      type: {
        type: String,
        enum: ["Point"],
        default: "Point",
      },
      coordinates: {
        type: [Number],
        required: true,
      },
      address: String,
      timestamp: Date,
    },
    checkInAddress: String,

    // ✅ Current location (updates in real-time)
    currentLocation: {
      type: {
        type: String,
        enum: ["Point"],
        default: "Point",
      },
      coordinates: {
        type: [Number],
      },
      address: String,
      timestamp: Date,
    },

    checkOutTime: Date,

    // ✅ Make checkOutLocation completely optional (no default)
    checkOutLocation: {
      type: {
        type: String,
        enum: ["Point"],
      },
      coordinates: {
        type: [Number],
      },
      address: String,
      timestamp: Date,
    },
    checkOutAddress: String,

    estimatedTimeToOffice: Number,
    distanceFromOffice: Number,
    totalHours: Number,
    status: {
      type: String,
      enum: ["CHECKED_IN", "REACHED_OFFICE", "CHECKED_OUT"],
      default: "CHECKED_IN",
    },
    reachedOfficeTime: Date,
  },
  { timestamps: true }
);

// Indexes for geospatial queries
AttendanceSchema.index({ checkInLocation: "2dsphere" });
AttendanceSchema.index({ currentLocation: "2dsphere" });
// Only index checkOutLocation if it exists
AttendanceSchema.index({ checkOutLocation: "2dsphere" }, { sparse: true });

module.exports = mongoose.model("Attendance", AttendanceSchema);
